s3parq/publish_parq.py

```diff
@@ -503,3 +538,105 @@ def publish(bucket: str, key: str, partitions: List[str], dataframe: pd.DataFram
+        dataframe (pd.DataFrame): Dataframe to be published
+        custom_redshift_columns (dict): 
+            This dictionary contains custom column data type definitions for redshift.
+            The params should be formatted as follows:
+                - column name (str)
+                - data type (str)
+        redshift_params (dict, Optional):
+            This dictionary should be provided in the following format in order
+            for data to be published to Spectrum. Leave out entirely to avoid
+            publishing to Spectrum.
+            The params should be formatted as follows:
+                - schema_name (str): Name of the Spectrum schema to publish to
+                - table_name (str): Name of the table to write the dataset as
+                - iam_role (str): Role to take while writing data to Spectrum
+                - region (str): AWS region for Spectrum
+                - cluster_id (str): Spectrum cluster id
+                - host (str): Redshift Spectrum host name
+                - port (str): Redshift Spectrum port to use
+                - db_name (str): Redshift Spectrum database name to use
+                - ec2_user (str): If on ec2, the user that should be used
+
+    Returns:
+        A str list of all the newly published object keys
+    """
+    logger.debug(
+        "Running custom publish...")
+
+    session_helper = None
+
+    if redshift_params:
+        if "index" in dataframe.columns:
+            raise ValueError(
+                "'index' is a reserved keyword in Redshift. Please remove or rename your DataFrame's 'index' column.")
+
+        logger.debug(
+            "Found redshift parameters. Checking validity of params...")
+        redshift_params = validate_redshift_params(redshift_params)
+        logger.debug("Redshift parameters valid. Opening Session helper.")
+        session_helper = SessionHelper(
+            region=redshift_params['region'],
+            cluster_id=redshift_params['cluster_id'],
+            host=redshift_params['host'],
+            port=redshift_params['port'],
+            db_name=redshift_params['db_name'],
+            ec2_user=redshift_params['ec2_user']
+        )
+
+        session_helper.configure_session_helper()
+        publish_redshift.create_schema(
+            redshift_params['schema_name'], redshift_params['db_name'], redshift_params['iam_role'], session_helper)
+        logger.debug(
+            f"Schema {redshift_params['schema_name']} created. Creating table {redshift_params['table_name']}...")
+
+        publish_redshift.create_custom_table(
+            redshift_params['table_name'], redshift_params['schema_name'], partitions, s3_url(bucket, key), custom_redshift_columns, session_helper)
+        logger.debug(f"Custom table {redshift_params['table_name']} created.")
+
+    logger.debug("Checking publish params...")
+    check_empty_dataframe(dataframe)
```

Comment:
Mostly a comment for later's sake but can't believe I left the check after the RedShift stuff already ran 🤦‍♀️keep for consistency but we should probably fix that later

Reply:
No worries!